File: 2141-Maximum Running Time of N Computers.py
class Solution(object):
    def maxRunTime(self, n, batteries):
        """
        :type n: int
        :type batteries: List[int]
        :rtype: int
        """
        left, right = min(batteries), sum(batteries)//n

        def checkOK(targetTime):
            totTime = 0
            totTarget = n*targetTime
            for battery in batteries:
                if battery > targetTime:
                    totTime += targetTime
                else:
                    totTime += battery
                if totTime >= totTarget:
                    return True
            return False
        
        while left + 1 < right:
            mid = left + (right-left)//2
            if checkOK(mid):
                left = mid
            else:
                right = mid
        
        if checkOK(right):
            return right
        else:
            return left

# Sol 2

        batteries.sort()
        arr = batteries[len(batteries)-n:]
        excess = sum(batteries[:len(batteries)-n])
        
        for i in range(1, len(arr)):
            diff = arr[i] - arr[i-1]
            if excess >= diff*i:
                excess -= diff*i
            else:
                return arr[i-1] + excess//i
        
        return arr[-1] + excess//n

File: C++/2141-Maximum Running Time of N Computers.cpp
class Solution {
public:
    bool canReach(long long time, int n, vector<int>& batteries) {
        long long batteryTime = 0;
        for (long long b:batteries) {
            batteryTime += min(b, time);
        }

        return batteryTime >= time*n;
    }
    long long maxRunTime(int n, vector<int>& batteries) {
        long long left = *min_element(begin(batteries), end(batteries));
        long long tot = accumulate(begin(batteries), end(batteries), 0LL);  // use 0LL, or we'll have runtime error
        long long right = tot/n;

        while (left + 1 < right) {
            long long mid = left + (right-left)/2;
            if (canReach(mid, n, batteries)) {
                left = mid;
            } else {
                right = mid;
            }
        }

        if (canReach(right, n, batteries)) {
            return right;
        }
        return left;
    }
};